Firestore/core/src/firebase/firestore/model/field_value.h

```diff
@@ -78,6 +78,8 @@ class FieldValue {
     // position instead, see the doc comment above.
   };
 
+  using ObjectT = std::map<std::string, FieldValue>;
```

Comment:
I'm not a super-huge fan of this name. It's the third different way of naming these "contents of a field value" subtypes.

We've got:
  * ServerTimestamp
  * ReferenceValue
  * FieldValue::ObjectT

I was going to write a whole treatise on the "T" suffix too but I don't think it really applies here.

I suggest adopting a convention for these: make them something like FooContents for each complex Foo that has a FooValue factory method.

That suggests this should be:

```c++
using ObjectContents = std::map<std::string, FieldValue>;
```

... and that we should fix the other content-of-FieldValue types to conform.

I'm actually okay with nearly any convention like this that compiles, but I'm not okay with three types and three different ways of naming and nesting them.

Reply:
I see you're point, but I'm not wild about the 'XContents' convention. The map is the object itself, not just the contents. (Or in other words, the methods like `string_value()` return a `string`, not a `StringContents`.)

How about, we drop the 'T', drop the 'Value' and just name them what they are:

Option1:
* FieldValue::ServerTimestamp
* FieldValue::Reference
* FieldValue::Object

Or alternatively, add 'Value' to everything (though I think that may be redundant... OTOH, the methods are explicitly named `something_value()` and this is also consistent with obj-c... but inconsistent with `primitive-ish-type_value()`):

Option2:
* FieldValue::ServerTimestampValue
* FieldValue::ReferenceValue
* FieldValue::ObjectValue

And of course, both of the above imply moving the definition into FieldValue. Alternatively, we could leave them outside the class:

Option3[4]:
* f::f::model::ServerTimestamp[Value]
* f::f::model::Reference[Value]
* f::f::model::Object[Value]

Or your original proposal [also with moving the defns into FieldValue, for completeness]

Option5[6]:
* [FieldValue::]ServerTimstampContents
* [FieldValue::]ReferenceContents
* [FieldValue::]ObjectContents

Thoughts?

Reply:
The C++ `FieldValue` in its current state is fairly inconsistent with the other platforms so I'm not sure we should drive decisions based on what it does now (e.g. `string_value()` returning a string). Our design waypoint should be the other platforms.

On Android, Typescript, and the current Objective-C client FooValue is a subclass of FieldValue modeling some Firestore-specific type Foo. For each Foo there's an underlying representation type that FooValue wraps to implement Firestore-specific semantics for equality, ordering, and mutation.

So the `std::string` of a `StringValue` and the `map` of the `ObjectValue` are representation types. As far as most of the rest of the client is concerned the `FooValue` is the thing of interest. There are two exceptions:

  * the _user_ view of the data from a snapshot (the `DocumentSnapshot.data()` method), and
  * the _network_ view in the serializer.

In neither case is `std::map<std::string, FieldValue>` directly the thing of interest. It's useful for rendering those other views only because we don't have a rendering abstraction, but this is only the implementation leaking out.

So this is the basis of my comments. I disagree that `std::map<std::string, FieldValue>` "is the Object". It's the internal representation of an Object. "Contents" was my attempt to give a class of names to the representation types.

What's confusing is that the `std::string` of `StringValue` is all of these: it's the user view, network view, and the representation type.

Once upon a time on the other platforms `FieldValue.value()` rendered externally visible snapshots but that's no longer the case. In the _user_ case, rendering Objects is now handled by [`DocumentSnapshot.convertObject`](http://google3/java/com/google/android/gmscore/dev/thick_client/firebase-firestore/src/com/google/firebase/firestore/DocumentSnapshot.java?l=402&rcl=190142681). In the _network_ case, it's handled by [`RemoteSerializer.encodeMapValue`](http://google3/java/com/google/android/gmscore/dev/thick_client/firebase-firestore/src/com/google/firebase/firestore/remote/RemoteSerializer.java?l=339&rcl=185935790). Both reach through to the Object's internal representation. AFAICT `ObjectValue.value()` is dead code that exists to satisfy a contract that doesn't apply anymore.

In some ways the C++ implementation cheats and conflates some of these concepts which makes this harder to reason about than it needs to be.
  * There are no `FooValue` types
  * The `foo_value()` method seems roughly equivalent to `(Foo) field.value()` on Android, but as discussed above this isn't actually useful

We could redefine `foo_value()` return the representation type of the `FieldValue` but it seems like the name is lying a bit. Perhaps in the short term this is acceptable though I think longer term we should work toward bringing these things back into parity with the other platforms so we don't confuse ourselves so much.

So in my ideal world we would have an `ObjectValue` type that would contain a `std::map<std::string, FieldValue>` and `object_value()` would serve as a kind of a cast more than anything else. The type you've named here is `ObjectValue::Map`, `ObjectValue::Rep`, or `ObjectValue::Contents` or something like that.

Outside the scope of this PR we should probably push C++ FieldValue more towards the other platforms and make the FooValue types explicit, and make the `FieldValue` union type more just a means of doing dispatch rather than having any of the implementation details in it.

Reply:
Ok; as discussed, we're using ObjectValue::Map, with a new ObjectValue representative type, which for now is just a struct with a map in it. PTAL.

Reply:
`ObjectT` is still here